### ISSUE TITLE
Concatenating 1D matrices to 2D matrices

### ISSUE THREAD

**Issue by ashwinmr** [josdejong/mathjs#1148]
I am using mathjs for some machine learning algorithms. I'm familiar with matlab where it is very easy to build matrices by concatenating scalars, arrays and matrices.
What's the best way to do this in mathjs ?

Example use case:
```
a = math.matrix([[1,2],[3,4]])
b = math.matrix([5,6])
math.concat(b,a,1)
```
Desired result:
```
[[5,1,2]
 [6,3,4]]
```
However, this results in a dimension error.

I could transpose `b` before concatenation. But it does not work since it does not make it 2D.

I am currently explicitly converting `b` to 2D using `b.reshape([2,1])`, but this requires keeping track of all outputs (sometimes from functions like `math.range()` and making sure to convert them to 2D which requires also computing their size.

Another option would be to force all matrices to be at least 2 dimensional. Is there a way to do that ?

Am I overlooking some easier solution ?

**Comment by ashwinmr** [comment]
I'm using this helper function to make all matrices at least 2 dimensional for now
```
/**
 * Make a number, array or matrix 2 or more dimensional
 * If input is already 2 or more dimentional it is returned as is
 * Arrays are columns by default
 * @param {number|array|matrix} a - input number, array or matrix
 * @returns {matrix} - a 2 or more d matrix
 */
function Make_Mat(a) {

    // Make input into matrix
    // Check if input is a number
    if (typeof a === 'number') {
        a = math.matrix([a])
    } else {
        a = math.matrix(a)
    }

    // Get number of dimensions
    let d = math.matrix(a.size()).size()[0]

    // Check if already 2D or greater
    if (d >= 2) {
        return a
    } else {
        let n = a.size()[0]

        // Check if empty matrix
        if (n === 0) {
            return a.reshape([0, 0])
        } else {
            return a.reshape([n, 1])
        }
    }
}
```

**Comment by josdejong** [comment]
Thanks for sharing. Would be nice if we could make `concat` more powerful so that it can automatically merge vectors with matrices like these util functions do.

Anyone interested in improving `concat`?

**Comment by ashwinmr** [comment]
While using my helper functions, I noticed that the `reshape` function` is very slow. It took 20 seconds to transpose a 1x100000 matrix. I'm guessing this is expected.

I tried the same operation using a matrix (already created with 2 dimensions) and it took around 40 milliseconds. So I think my main issue is that I am using the wrong method to convert an array into a matrix. What's the best way to do that ?

**Comment by josdejong** [comment]
You create a Matrix from an Array using `var m = math.matrix(a)` like you already use in your code example.

I'm not sure why `reshape` on an Arrow would be so much slower than on a Matrix, that sounds weird. Can you share a code example on how to reproduce this?

**Comment by ashwinmr** [comment]
Here's what I'm doing:
```
var a = math.matrix([1,2,3,4])
```
This results in a 1 dimensional matrix. When I try to transpose it, it stays 1 dimensional. So I need to somehow make it 2 dimensional before I am allowed to concatenate it with anything.
If I try what you suggested:
```
var b = math.matrix(a)
```
`b` is again 1 dimensional.
The only solution I found was to use the `reshape` function.
```
var b = a.reshape([4,1]) 
```
Now `b` is 2 dimensional and can be concatenated with a matrix.

However, this is when I realized how slow the `reshape` function is. I didn't dig too deep, but it is a recursive function, which might be the reason.

Try this code snippet
```
var a = math.ones(1, 100000)

console.time('transpose')
var b = math.transpose(a)
console.timeEnd('transpose')

console.time('reshape')
var c = a.reshape([100000, 1])
console.timeEnd('reshape')
````

both `b` and `c` end up as the same matrix. But here is my console output:
```
transpose: 45.275ms
reshape: 5825.299ms
```
